zebra-chain/src/orchard/action.rs

```diff
@@ -0,0 +1,69 @@
+use std::io;
+
+use halo2::pasta::pallas;
+
+use crate::{
+    primitives::redpallas::{self, SpendAuth},
+    serialization::{
+        serde_helpers, ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize,
+    },
+};
+
+use super::{
+    commitment::{self, ValueCommitment},
+    keys,
+    note::{self, Nullifier},
+};
+
+/// An Action description, as described in the [Zcash specification §7.3][actiondesc].
+///
+/// Action transfers can optionally perform a spend, and optionally perform an
+/// output.  Action descriptions are data included in a transaction that
+/// describe Action transfers.
+///
+/// [actiondesc]: https://zips.z.cash/protocol/nu5.pdf#actiondesc
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Action {
+    /// A value commitment to net value of the input note minus the output note
+    pub cv: commitment::ValueCommitment,
+    /// The nullifier of the input note being spent.
+    pub nullifier: note::Nullifier,
+    /// The randomized validating key for spendAuthSig,
+    pub rk: redpallas::VerificationKeyBytes<SpendAuth>,
+    /// The x-coordinate of the note commitment for the output note.
+    #[serde(with = "serde_helpers::Base")]
+    pub cm_x: pallas::Base,
+    /// An encoding of an ephemeral Pallas public key.
+    pub ephemeral_key: keys::EphemeralPublicKey,
+    /// A ciphertext component for the encrypted output note.
+    pub enc_ciphertext: note::EncryptedNote,
+    /// A ciphertext component for the encrypted output note.
+    pub out_ciphertext: note::WrappedNoteKey,
```

Comment:
We can be more precise than the spec here:
```suggestion
    /// A ciphertext component for the encrypted output note.
    pub enc_ciphertext: note::EncryptedNote,
    /// A ciphertext component that allows the holder of a full viewing key to
    /// recover the recipient diversified transmission key and the ephemeral
    /// private key (and therefore the entire note plaintext)
    pub out_ciphertext: note::WrappedNoteKey,
```

Reply:
(This is a copy of an earlier definition of `enc_ciphertext` from the spec, I've also submitted a PR to copy it into the serialization table that corresponds to this data structure.)